### PR TITLE
extract build/apply functions about snapshot

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

## What have you changed? (mandatory)
Extract build/apply functions from SnapManager. These works can be done out of SnapManger, so let's extract them so that we can simplify SnapManager more easily.

## What are the type of the changes? (mandatory)
Engineering.

## How has this PR been tested? (mandatory)
Unit tests.

### CODE REVIEW

**Comment by hicqu** [comment]
@Connor1996 @zhangjinpeng1987 PTAL, thanks!

**Comment by hicqu** [comment]
PTAL @zhangjinpeng1987 @overvenus

File: src/raftstore/store/snap/io.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    end_key: &[u8],
+    io_limiter: Option<&IOLimiter>,
+) -> Result<BuildStatistics, Error> {
+    let mut sst_writer = create_sst_file_writer(snap, cf, path)?;
+    let mut stats = BuildStatistics::default();
+    let mut base = if io_limiter.is_some() { 0 } else { usize::MAX };
+    box_try!(snap.scan_cf(cf, start_key, end_key, false, |key, value| {
+        if let Some(ref io_limiter) = io_limiter {
+            if stats.total_size >= base {
+                let next_request = io_limiter.get_max_bytes_per_time();
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap/io.rs]: The logic between L73~L78 is not consistent with master.

File: src/raftstore/store/snap/io.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        stats.total_size += key.len() + value.len();
+        if let Err(e) = sst_writer.put(key, value) {
+            let io_error = io::Error::new(io::ErrorKind::Other, e);
+            return Err(io_error.into());
+        }
+        Ok(true)
+    }));
+    if stats.key_count > 0 {
+        box_try!(sst_writer.finish());
+        drop(sst_writer);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap/io.rs]: Don't need to drop sst_writer explicitly.

File: src/raftstore/store/snap/io.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    let mut decoder = BufReader::new(box_try!(File::open(path)));
+    let cf_handle = box_try!(get_cf_handle(&db, cf));
+    let wb = WriteBatch::new();
+    loop {
+        if stale_detector.is_stale() {
+            return Err(Error::Abort);
+        }
+        let key = box_try!(decoder.decode_compact_bytes());
+        if key.is_empty() {
+            box_try!(db.write(&wb));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap/io.rs]: Only need to write db when writebatch.data_size() > 0.

File: src/raftstore/store/snap/io.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            return Ok(());
+        }
+        let value = box_try!(decoder.decode_compact_bytes());
+        box_try!(wb.put_cf(cf_handle, &key, &value));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap/io.rs]: Plese check key in region first.

Comment (benpigchu) [reply src/raftstore/store/snap/io.rs]: Seems the key in region check do not exist even in the master branch.

File: src/raftstore/store/snap/io.rs [hunk 5]
```diff
@@ -68,17 +68,22 @@ pub fn build_sst_cf_file(
...
+        .map_or(0 as i64, |l| l.get_max_bytes_per_time());
+    let mut bytes: i64 = 0;
     box_try!(snap.scan_cf(cf, start_key, end_key, false, |key, value| {
+        let entry_len = key.len() + value.len();
         if let Some(ref io_limiter) = io_limiter {
-            if stats.total_size >= base {
-                let next_request = io_limiter.get_max_bytes_per_time();
-                io_limiter.request(next_request);
-                base += next_request as usize;
+            // seems the usage of io_limiter here do not match its design
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap/io.rs]: How should we use it?

File: src/raftstore/store/snap.rs [hunk 6]
```diff
@@ -698,135 +626,55 @@ impl Snap {
...
+                )?
             };
-            snap_key_count += cf_key_count;
+            cf_file.kv_count = cf_stat.key_count as u64;
+            if cf_file.kv_count > 0 {
+                fs::rename(&cf_file.tmp_path, &cf_file.path)?;
+                cf_file.checksum = calc_crc32(&cf_file.path)?;
+                cf_file.size = get_file_size(&cf_file.path)?;
+                self.size_track.fetch_add(cf_file.size, Ordering::SeqCst);
+            }
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap.rs]: Else should delete the tmp file.

File: src/raftstore/store/snap/io.rs (hunk 5 above)
Comment (benpigchu) [reply src/raftstore/store/snap/io.rs]: I think it should be something like `while(bytes>=base){bytes-=base;...}`

File: src/raftstore/store/snap.rs [hunk 7]
```diff
@@ -698,135 +626,55 @@ impl Snap {
...
-                })?;
-                (key_count, size)
+                let limiter = self.limiter.as_ref().map(|c| c.as_ref());
+                snap_io::build_sst_cf_file(
+                    path, kv_snap, cf_file.cf, &begin_key, &end_key, limiter,
+                )?
             };
-            snap_key_count += cf_key_count;
+            cf_file.kv_count = cf_stat.key_count as u64;
+            if cf_file.kv_count > 0 {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap.rs]: Please migrate the comment why we use kv_count to decide if the file is empty either than use the file size.

**Pull request merged**